libbeat/autodiscover/providers/docker/docker.go

```diff
@@ -119,41 +121,87 @@ func (d *Provider) Start() {
 	}()
 }
 
-func (d *Provider) emitContainer(event bus.Event, flag string) {
+type dockerMetadata struct {
+	// Old selectors [Deprecated]
+	Docker common.MapStr
+
+	// New ECS-based selectors
+	Container common.MapStr
+
+	// Metadata used to enrich events, like ECS-based selectors but can
+	// have modifications like dedotting
+	Metadata common.MapStr
+}
+
+func (d *Provider) generateMetaDocker(event bus.Event) (*docker.Container, *dockerMetadata) {
 	container, ok := event["container"].(*docker.Container)
 	if !ok {
-		logp.Err("Couldn't get a container from watcher event")
-		return
+		logp.Error(errors.New("Couldn't get a container from watcher event"))
+		return nil, nil
 	}
 
-	var host string
-	if len(container.IPAddresses) > 0 {
-		host = container.IPAddresses[0]
-	}
+	// Don't dedot selectors, dedot only metadata used for events enrichment
 	labelMap := common.MapStr{}
+	metaLabelMap := common.MapStr{}
 	for k, v := range container.Labels {
 		safemapstr.Put(labelMap, k, v)
+		if d.config.Dedot {
+			label := common.DeDot(k)
+			metaLabelMap.Put(label, v)
+		} else {
+			safemapstr.Put(metaLabelMap, k, v)
+		}
 	}
 
-	meta := common.MapStr{
-		"container": common.MapStr{
-			"id":     container.ID,
-			"name":   container.Name,
-			"image":  container.Image,
+	meta := &dockerMetadata{
+		Docker: common.MapStr{
+			"container": common.MapStr{
+				"id":     container.ID,
+				"name":   container.Name,
+				"image":  container.Image,
+				"labels": labelMap,
+			},
+		},
+		Container: common.MapStr{
+			"id":   container.ID,
+			"name": container.Name,
+			"image": common.MapStr{
+				"name": container.Image,
+			},
 			"labels": labelMap,
 		},
+		Metadata: common.MapStr{
+			"docker": common.MapStr{
+				"container": common.MapStr{
+					"id":     container.ID,
+					"name":   container.Name,
+					"image":  container.Image,
+					"labels": metaLabelMap,
+				},
+			},
+		},
 	}
+
+	return container, meta
+}
+
+func (d *Provider) emitContainer(event bus.Event, flag string) {
+	container, meta := d.generateMetaDocker(event)
+	var host string
+	if len(container.IPAddresses) > 0 {
```

Comment:
Looks like if the condition here is hit, we'll return `nil` for `container` and this will panic:
https://github.com/elastic/beats/pull/10899/files#diff-fc73524502be96120bf7d211699c2657R138

Reply:
Good catch, this also affects 7.0 branch that I already merged :slightly_frowning_face: I will open another PR to fix it in master.

Reply:
Fixed here and opened PR for master/7.0.